### PR TITLE
Add config options for subject prefixes

### PR DESCRIPTION
**Pull request by tpayen**
Some french mail clients use "Tr:" instead of "Fwd:" as forward indicator in subject. 
I don't know if there is such exceptions in others languages in which case this maybe worth the use of a configuration parameter instead of an hardcoded regex ?

### CODE REVIEW

**Comment by alecpl** [comment]
We have an old ticket for this: #4981. I think it should be a config option with a default that contains most common cases.

**Comment by tpayen** [comment]
Hi Alec,
I've added a new commit to support config options for subject prefixes.
4 more options are possible : 
 - default_forward_prefix (default = 'Fwd:') : the default prefix when forwarding an email
 - default_response_prefix (default = 'Re:') : the default prefix when responding to an email
 - subject_response_prefixes (default = ['Re:']) : the prefixes list to clean for responses
 - subject_forward_prefixes (default = ['Fwd:', 'Fw:']) : the prefixes list to clean for forwarded messages

File: program/include/rcmail_sendmail.php [hunk 1]
```diff
@@ -1202,7 +1207,7 @@ public static function reply_subject($subject)
         // replace (was: ...) (#1489375)
         $subject = preg_replace('/\s*\([wW]as:[^\)]+\)\s*$/', '', $subject);
 
-        return 'Re: ' . $subject;
+        return rcmail::get_instance()->config->get('default_response_prefix', 'Re:') . ' ' . $subject;
```
Comment (alecpl) [inline program/include/rcmail_sendmail.php:1210]: I would remove `default_` from these option names. Also, what if it is set to an empty string? You should trim the return value.

File: program/lib/Roundcube/rcube_imap_generic.php [hunk 2]
```diff
@@ -2742,7 +2742,10 @@ public static function sortHeaders($messages, $field, $order = 'ASC')
                     $value = str_replace('"', '', $value);
 
                     if ($field == 'subject') {
-                        $value = preg_replace('/^(Re:\s*|Fwd:\s*|Fw:\s*)+/i', '', $value);
+                        $forward_prefixes = rcmail::get_instance()->config->get('subject_forward_prefixes', ['Fwd:', 'Fw:']);
+                        $response_prefixes = rcmail::get_instance()->config->get('subject_response_prefixes', ['Re:']);
+                        $data = array_map(function($prefix) { return $prefix . '\s*'; }, array_merge($forward_prefixes, $response_prefixes));
+                        $value = preg_replace('/^('.implode('|', $data).')+/i', '', $value);
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_imap_generic.php:2748]: So, we have similar code in a few places. I would suggest to unify and de-duplicate this.

I.e. create `rcube_utils::remove_subject_prefix($subject, $mode)` where mode would be 'reply', 'forward' or 'both'.

**Review by alecpl — CHANGES_REQUESTED** [review]

**Comment by tpayen** [comment]
Thanks for these relevant remarks. I've added a new commit to add the `rcube_utils::remove_subject_prefix` function and make everything a little cleaner

File: program/include/rcmail_sendmail.php [hunk 3]
```diff
@@ -1192,17 +1192,14 @@ public static function reply_subject($subject)
...
-        $prefix = '/^(re:|re\[\d\]:|re-\d:)\s*/i';
-        do {
-            $subject = preg_replace($prefix, '', $subject, -1, $count);
-        }
-        while ($count);
+        //  Add config options for subject prefixes (#7929) 
+        $subject = rcube_utils::remove_subject_prefix($subject, 'reply');
 
         // replace (was: ...) (#1489375)
         $subject = preg_replace('/\s*\([wW]as:[^\)]+\)\s*$/', '', $subject);
```
Comment (alecpl) [inline program/include/rcmail_sendmail.php:1199]: How about moving this into remove_subject_prefix() (in 'reply' mode)?

File: program/lib/Roundcube/rcube_utils.php [hunk 4]
```diff
@@ -1580,4 +1580,41 @@ public static function temp_filename($file_name, $unique = true, $create = true)
...
+
+    /**
+     * Clean the subject from reply and forward prefix
+     * 
+     * @param string $subject Subject to clean
+     * @param string $mode Mode of cleaning : reply, forward or both
+     * 
+     * @return string Cleaned subject
+     */
+    public static function remove_subject_prefix($subject, $mode = 'both')
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_utils.php:1592]: How do you feel about adding some tests in tests/Framework/Utils.php?

File: program/lib/Roundcube/rcube_utils.php [hunk 5]
```diff
@@ -1580,4 +1580,41 @@ public static function temp_filename($file_name, $unique = true, $create = true)
...
+    {
+        // Clean subject prefix for reply, forward or both
+        if ($mode == 'both') {
+            $reply_prefixes = rcmail::get_instance()->config->get('subject_reply_prefixes', ['Re:']);
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_utils.php:1596]: Let's do `$config = rcmail::get_instance()->config` first and then use it to make the code shorter.

File: program/lib/Roundcube/rcube_utils.php [hunk 6]
```diff
@@ -1580,4 +1580,41 @@ public static function temp_filename($file_name, $unique = true, $create = true)
...
+            $prefixes = rcmail::get_instance()->config->get('subject_reply_prefixes', ['Re:']);
+        }
+        else if ($mode == 'forward') {
+            $prefixes = rcmail::get_instance()->config->get('subject_forward_prefixes', ['Fwd:', 'Fw:']);
+        }
+
+        // replace Re:, Re[x]:, Re-x (#1490497)
+        $pieces = array_map(function($prefix) {
+            $prefix = strtolower(str_replace(':', '', $prefix)); 
+            return $prefix.':|'.$prefix.'\[\d\]:|'.$prefix.'-\d:';
```
Comment (alecpl) [inline program/lib/Roundcube/rcube_utils.php:1610]: Add a space between dots, please. Or everything using double-quote'd syntax.

**Review by alecpl — CHANGES_REQUESTED** [review]

File: program/lib/Roundcube/rcube_utils.php (hunk 4 above)
Comment (tpayen) [reply program/lib/Roundcube/rcube_utils.php:1592]: I'll try but I don't promise anything  ;-)
I'll do an independent commit for this

**Pull request merged**